File: Source/Game/Items/Relics/Relic.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;

public abstract class Relic
{
    private static trigger CanBuyRelicsTrigger;
    private static List<player> CanBuyRelics;

    public static int RequiredLevel { get; } = 12;
    public static int RelicIncrease { get; } = 16;
    public static int RelicSellLevel { get; } = 15;
    public static int MaxRelics { get; } = 2;
    public string Name { get; }
    public string Description { get; }
    public int ItemID { get; }
    public int Cost { get; }
    public string IconPath { get; }
    public int UpgradeLevel { get; private set; } = 0;
    public int MaxUpgradeLevel => Upgrades.Count;
    public int RelicAbilityID { get; protected set; }
    public List<RelicUpgrade> Upgrades { get; } = new List<RelicUpgrade>();

    public Relic(string name, string desc, int relicAbilityID, int itemID, int cost, string iconPath)
    {
        Name = name;
        Description = desc;
        RelicAbilityID = relicAbilityID;
        ItemID = itemID;
        Cost = cost;
        IconPath = iconPath;
    }

    public abstract void ApplyEffect(unit Unit);

    public abstract void RemoveEffect(unit Unit);

    public RelicUpgrade GetCurrentUpgrade()
    {
        if (Upgrades == null || Upgrades.Count == 0) return null;
        if (UpgradeLevel >= Upgrades.Count) return Upgrades[Upgrades.Count - 1];
        return Upgrades[UpgradeLevel];
    }

    public bool CanUpgrade(player player) => PlayerUpgrades.GetPlayerUpgrades(player).GetUpgradeLevel(GetType()) < MaxUpgradeLevel;

    public bool Upgrade(unit Unit)
    {
        if (!CanUpgrade(Unit.Owner)) return false;
        UpgradeLevel++;
        PlayerUpgrades.IncreaseUpgradeLevel(GetType(), Unit);
        SetUpgradeLevelDesc(Unit);
        RemoveEffect(Unit);
        ApplyEffect(Unit);
        return true;
    }

    public static int GetRelicCountForLevel(int currentLevel)
    {
        var count = currentLevel - RelicIncrease + 1; // account for level 10 relic ..
        return count < 0 ? 0 : count >= MaxRelics ? MaxRelics : count;
    }

    public void SetUpgradeLevelDesc(unit Unit)
    {
        var upgradeLevel = PlayerUpgrades.GetPlayerUpgrades(Unit.Owner).GetUpgradeLevel(GetType());
        if (upgradeLevel == 0) return;

        var item = Utility.UnitGetItem(Unit, ItemID);
        if (item == null) return;

        var tempName = item.Name;
        var newUpgradeText = $"{Colors.COLOR_TURQUOISE}[Upgrade: {upgradeLevel}]{Colors.COLOR_RESET}";

        if (tempName.StartsWith($"{Colors.COLOR_TURQUOISE}[Upgrade:"))
        {
            var endIndex = tempName.IndexOf("]|r") + 3;
            tempName = tempName.Substring(endIndex).Trim();
        }

        item.Name = $"{newUpgradeText} {tempName}";
    }

    public static void RegisterRelicEnabler()
    {
        // Ability to Purchase Relics
        PlayerUpgrades.Initialize();
        CanBuyRelicsTrigger ??= trigger.Create();
        CanBuyRelics ??= new List<player>();
        Blizzard.TriggerRegisterAnyUnitEventBJ(CanBuyRelicsTrigger, playerunitevent.HeroLevel);
        CanBuyRelicsTrigger.AddAction(() =>
        {
            try
            {
                if (CanBuyRelics.Contains(@event.Unit.Owner)) return;
                if (@event.Unit.HeroLevel < RequiredLevel) return;
                CanBuyRelics.Add(@event.Unit.Owner);
                RelicUtil.EnableRelicBook(@event.Unit);
                RelicUtil.DisableRelicAbilities(@event.Unit);
                ProtectionOfAncients.SetProtectionOfAncientsLevel(@event.Unit);
                @event.Unit.Owner.DisplayTimedTextTo(4.0f, $"{Colors.COLOR_TURQUOISE}You may now buy relics from the shop!{Colors.COLOR_RESET}");
            }
            catch (Exception e)
            {
                Logger.Warning($"Error in RegisterLevelTenTrigger {e.Message}");
            }
        });
    }
}

File: src/Game/Items/Relics/Relic.ts
import { Logger } from 'src/Events/Logger/Logger'
import { ProtectionOfAncients } from 'src/Game/ProtectionOfAncients'
import { Colors } from 'src/Utility/Colors/Colors'
import { Utility } from 'src/Utility/Utility'
import { getTriggerUnit } from 'src/Utility/w3tsUtils'
import { MapPlayer, Trigger, Unit } from 'w3ts'
import { PlayerUpgrades } from './PlayerUpgrades'
import { RelicUpgrade } from './RelicUpgrade'
import { RelicUtil } from './RelicUtil'

export abstract class Relic {
    private static CanBuyRelicsTrigger: Trigger
    private static CanBuyRelics: MapPlayer[]

    public static RequiredLevel = 12
    public static RelicIncrease = 16
    public static RelicSellLevel = 15
    public static MaxRelics = 2
    public name: string
    public Description: string
    public ItemID = 0
    public Cost = 0
    public IconPath: string
    public UpgradeLevel = 0
    public RelicAbilityID = 0
    public Upgrades: RelicUpgrade[] = []

    public constructor(
        name: string,
        desc: string,
        relicAbilityID: number,
        itemID: number,
        cost: number,
        iconPath: string
    ) {
        this.name = name
        this.Description = desc
        this.RelicAbilityID = relicAbilityID
        this.ItemID = itemID
        this.Cost = cost
        this.IconPath = iconPath
    }

    public abstract ApplyEffect(Unit: Unit): void

    public abstract RemoveEffect(Unit: Unit): void

    public GetCurrentUpgrade(): RelicUpgrade {
        if (this.Upgrades === null || this.Upgrades.length === 0) return null as never
        if (this.UpgradeLevel >= this.Upgrades.length) return this.Upgrades[this.Upgrades.length - 1]
        return this.Upgrades[this.UpgradeLevel]
    }

    public CanUpgrade(player: MapPlayer) {
        return PlayerUpgrades.GetPlayerUpgrades(player).GetUpgradeLevel(this.constructor.name) < this.Upgrades.length
    }

    public Upgrade(Unit: Unit) {
        if (!this.CanUpgrade(Unit.owner)) return false
        this.UpgradeLevel = this.UpgradeLevel + 1
        PlayerUpgrades.IncreaseUpgradeLevel(this.constructor.name, Unit)
        this.SetUpgradeLevelDesc(Unit)
        this.RemoveEffect(Unit)
        this.ApplyEffect(Unit)
        return true
    }

    public static GetRelicCountForLevel(currentLevel: number) {
        const count = currentLevel - Relic.RelicIncrease + 1 // account for level 10 relic ..
        return count < 0 ? 0 : count >= Relic.MaxRelics ? Relic.MaxRelics : count
    }

    public SetUpgradeLevelDesc(Unit: Unit) {
        const upgradeLevel = PlayerUpgrades.GetPlayerUpgrades(Unit.owner).GetUpgradeLevel(this.constructor.name)
        if (upgradeLevel === 0) return

        const item = Utility.UnitGetItem(Unit, this.ItemID)
        if (!item) return

        let tempName = item.name
        const newUpgradeText = `${Colors.COLOR_TURQUOISE}[Upgrade: ${upgradeLevel}]${Colors.COLOR_RESET}`

        if (tempName.startsWith(`${Colors.COLOR_TURQUOISE}[Upgrade:`)) {
            const endIndex = tempName.indexOf(']|r') + 3
            tempName = tempName.substring(endIndex).trim()
        }

        item.name = `${newUpgradeText} ${tempName}`
    }

    public static RegisterRelicEnabler = () => {
        // Ability to Purchase Relics
        PlayerUpgrades.Initialize()
        Relic.CanBuyRelicsTrigger ??= Trigger.create()!
        Relic.CanBuyRelics ??= []
        Relic.CanBuyRelicsTrigger.registerAnyUnitEvent(EVENT_PLAYER_HERO_LEVEL)
        Relic.CanBuyRelicsTrigger.addAction(() => {
            try {
                if (Relic.CanBuyRelics.includes(getTriggerUnit().owner)) return
                if (getTriggerUnit().getHeroLevel() < Relic.RequiredLevel) return
                Relic.CanBuyRelics.push(getTriggerUnit().owner)
                RelicUtil.EnableRelicBook(getTriggerUnit())
                RelicUtil.DisableRelicAbilities(getTriggerUnit())
                ProtectionOfAncients.SetProtectionOfAncientsLevel(getTriggerUnit())
                getTriggerUnit()
                    .getOwner()!
                    .DisplayTimedTextTo(
                        4.0,
                        `${Colors.COLOR_TURQUOISE}You may now buy relics from the shop!${Colors.COLOR_RESET}`
                    )
            } catch (e) {
                Logger.Warning(`Error in RegisterLevelTenTrigger: ${e}`)
            }
        })
    }
}
